rollup/fees/rollup_fee.go

```diff
@@ -173,9 +162,58 @@ func readGPOStorageSlots(addr common.Address, state StateDB) gpoState {
 	gpoState.l1BlobBaseFee = state.GetState(addr, rcfg.L1BlobBaseFeeSlot).Big()
 	gpoState.commitScalar = state.GetState(addr, rcfg.CommitScalarSlot).Big()
 	gpoState.blobScalar = state.GetState(addr, rcfg.BlobScalarSlot).Big()
+	gpoState.penaltyThreshold = state.GetState(addr, rcfg.PenaltyThresholdSlot).Big()
+	gpoState.penaltyFactor = state.GetState(addr, rcfg.PenaltyFactorSlot).Big()
 	return gpoState
 }
 
+// estimateTxCompressionRatio estimates the compression ratio for transaction data using da-codec
+// compression_ratio(tx) = size(tx) * PRECISION / size(zstd(tx))
+func estimateTxCompressionRatio(data []byte, blockNumber uint64, blockTime uint64, config *params.ChainConfig) *big.Int {
+	if len(data) == 0 {
+		return rcfg.Precision
+	}
+
+	// Compress data using da-codec
+	compressed, err := encoding.CompressScrollBatchBytes(data, blockNumber, blockTime, config)
+	if err != nil {
+		log.Error("Batch compression failed, using 1.0 compression ratio", "error", err, "data size", len(data), "data", common.Bytes2Hex(data))
+		return rcfg.Precision
+	}
```

Comment:
I feel uneasy about this. Though this should never fail, if it does, it seems better if the sequencer node rejects the transaction. This issue is a bit similar to when `tx.MarshalBinary()` fails. Shall we simply propagate the error here?

Reply:
Good suggestion. Added here: [bd7f658](https://github.com/scroll-tech/go-ethereum/pull/1196/commits/bd7f65867cca6519f5d1ac2d300882d2cd26d869).